src/utils/typed-data.ts

```diff
@@ -0,0 +1,90 @@
+ * Hashes arbitrary object, can be used to inline the aci hash to contract source code
+ */
+export function hashJson(data: unknown): Buffer {
+  return hash(canonicalize(data) ?? '');
+}
+
+// TODO: move this type to calldata library https://github.com/aeternity/aepp-calldata-js/issues/215
+// based on https://github.com/aeternity/aepp-calldata-js/blob/82b5a98f9b308482627da8d7484d213e9cf87151/src/AciTypeResolver.js#L129
+export type AciValue = 'void' | 'unit' | 'int' | 'bool' | 'string' | 'bits' | 'hash' | 'signature'
+| 'address' | 'contract_pubkey' | 'Chain.ttl' | 'Chain.ga_meta_tx' | 'Chain.paying_for_tx'
+| 'Chain.base_tx' | 'AENS.pointee' | 'AENS.name' | 'MCL_BLS12_381.fr' | 'MCL_BLS12_381.fp'
+| { 'Set.set': readonly [AciValue] }
+| { bytes: number }
+| { list: readonly [AciValue] }
+| { map: readonly [AciValue, AciValue] }
+| { tuple: readonly AciValue[] }
+| { record: ReadonlyArray<{ name: string; type: AciValue }> }
+| { variant: ReadonlyArray<{ [key: string]: readonly AciValue[] }> }
+| { option: readonly [AciValue] }
+| { oracle: readonly [AciValue, AciValue] }
+| { oracle_query: readonly [AciValue, AciValue] };
+
+export interface Domain {
+  name?: string;
+  version?: number;
+  networkId?: string;
+  contractAddress?: Encoded.ContractAddress;
+}
+
+// TODO: replace with api on calldata side https://github.com/aeternity/aepp-calldata-js/issues/216
+export function encodeFateValue(
+  value: unknown,
+  aci: AciValue,
+): Encoded.ContractBytearray {
+  const contractByteArrayEncoder = new ContractByteArrayEncoder();
+  const aciTypeResolver = new AciTypeResolver([]);
+  aciTypeResolver.isCustomType = () => false;
+  return contractByteArrayEncoder.encode(aciTypeResolver.resolveType(aci), value);
+}
+
+// TODO: replace with api on calldata side https://github.com/aeternity/aepp-calldata-js/issues/216
+export function decodeFateValue(
+  value: Encoded.ContractBytearray,
+  aci: AciValue,
+): Encoded.ContractBytearray {
+  const contractByteArrayEncoder = new ContractByteArrayEncoder();
+  const aciTypeResolver = new AciTypeResolver([]);
+  aciTypeResolver.isCustomType = () => false;
+  return contractByteArrayEncoder.decodeWithType(value, aciTypeResolver.resolveType(aci));
+}
+
+/**
+ * Hashes domain object, can be used to inline domain hash to contract source code
+ */
+export function hashDomain(domain: Domain): Buffer {
+  const domainAci = {
+    record: [{
+      name: 'name',
+      type: { option: ['string'] },
+    }, {
+      name: 'version',
+      type: { option: ['int'] },
+    }, {
+      name: 'networkId',
+      type: { option: ['string'] },
+    }, {
+      name: 'contractAddress',
+      type: { option: ['contract_pubkey'] },
+    }],
+  } as const;
+  return hash(decode(encodeFateValue(domain, domainAci)));
+}
+
+export function hashTypedData(
+  data: Encoded.ContractBytearray,
+  aci: AciValue,
+  domain: Domain,
+): Buffer {
+  return hash(concatBuffers([hashDomain(domain), hashJson(aci), hash(decode(data))]));
```

Comment:
Should prefix the calculated hash with a magic number? I see that Ethereum developers care to have not overlapped signing payloads for transactions and other messages, but I don't understand the reasoning. For example,

> The initial 0x19 byte is intended to ensure that the signed_data is not valid RLP

https://eips.ethereum.org/EIPS/eip-191

Reply:
Yes I think there should be a prefix, though maybe not 0x19 since that was chosen because `Ethereum Signed Message:` happens to be 0x19 bytes long...

Reply:
I propose to define 0x1a00 prefix, with the second byte as a signature type (0x00 for typed data). E.g. 0x1a61 would be a prefix of a signed message the same as in EIP-191. This won't overlap with transaction signatures because they are prefixed with network id that can't contain ASCII control characters (0x1a, substitute).

Reply:
I propose to define 0x1a00 prefix, with the second byte as a signature type (0x00 for typed data). E.g. 0x1a61 would be a prefix of a signed message the same as in EIP-191. This won't overlap with transaction signatures because they are prefixed with network id that can't contain ASCII control characters (0x1a, substitute).